src/lib/work_partitioner/pending_zkapp_command.ml

```diff
@@ -65,19 +86,48 @@ let next_segment (t : t) =
   t.proofs_in_flight <- t.proofs_in_flight + 1 ;
   segment
 
-let next_subzkapp_job_spec (t : t) : Spec.Sub_zkapp.Stable.Latest.t option =
+let next_subzkapp_job_spec (t : t) :
+    (Spec.Sub_zkapp.Stable.Latest.t * Range.t) option =
   match next_merge t with Some _ as ret -> ret | None -> next_segment t
 
-let submit_proof (t : t) ~(proof : Ledger_proof.t)
-    ~(elapsed : Time.Stable.Span.V1.t) =
-  Deque.enqueue_back t.pending_mergeable_proofs proof ;
-  t.proofs_in_flight <- t.proofs_in_flight - 1 ;
-  t.elapsed <- Time.Span.(t.elapsed + elapsed)
+let submit_proof ~(proof : Ledger_proof.t) ~(elapsed : Time.Stable.Span.V1.t)
+    ~range:(Range.{ first; last } as range) (t : t) =
+  let should_accept =
+    first <= last
+    && RangeMap.closest_key t.pending_mergeable_proofs `Greater_or_equal_to
+         Range.{ first = last; last }
+       |> Option.map ~f:(fun (Range.{ first = next_first; _ }, _) ->
+              last < next_first )
+       |> Option.value ~default:true
+    && RangeMap.closest_key t.pending_mergeable_proofs `Less_or_equal_to
+         Range.{ first; last = first }
+       |> Option.map ~f:(fun (Range.{ last = previous_last; _ }, _) ->
+              previous_last < first )
+       |> Option.value ~default:true
+  in
+  if should_accept then (
+    t.pending_mergeable_proofs <-
+      RangeMap.add_exn ~key:range ~data:proof t.pending_mergeable_proofs ;
+    t.proofs_in_flight <- t.proofs_in_flight - 1 ;
+    t.elapsed <- Time.Span.(t.elapsed + elapsed) ;
+    Ok () )
+  else
+    let msg =
+      Printf.sprintf
+        "Pending zkapp command has a submission of range [%d, %d], that \
+         doesn't fit"
+        first last
+    in
+    Error (Error.of_string msg)
 
 let try_finalize (t : t) =
   if
     t.proofs_in_flight = 0
     && Queue.is_empty t.unscheduled_segments
-    && Deque.length t.pending_mergeable_proofs = 1
-  then Some (t.job, Deque.dequeue_back_exn t.pending_mergeable_proofs, t.elapsed)
+    && RangeMap.length t.pending_mergeable_proofs = 1
```

Comment:
I think I can check additionally if it correspond to the whole range of segments, unsure if it's needed